### DISCUSSION TITLE
Sharing about libraries used in elastalert2

### DISCUSSION THREAD

**Discussion by nsano-rururu** [jertel/elastalert2#1247] [Show and tell]
| name                    | python 3.12                                                                 |
| ----------------------- | --------------------------------------------------------------------------- |
| apscheduler             | △(4.0.0a3)                                                                          |
| aws-requests-auth       | ?                                                                           |
| boto3                   | 〇                                              |
| cffi                    | 〇(1.16.0)                                                                           |
| croniter                | ?                                                                           |
| elasticsearch           | ?                                                                           |
| envparse                | ?                                                                           |
| exotel                  | ?                                                                           |
| Jinja2                  | 〇<br>Test on Python 3.12-dev.                                              |
| jira                    | ?                                                                           |
| jsonpointer             | 〇                                                                           |
| jsonschema              | 〇                                                                           |
| prison                  | ?                                                                           |
| prometheus_client       | 〇(0.18.0～)                                                                           |
| python-dateutil         | ?|
| PyYAML                  |  ?                            |
| py-zabbix               | ?                                                                           |
| requests                | 〇                                              |
| sortedcontainers        | △(Waiting for pull request merge)                            |
| statsd-tags             | ?                                                                           |
| stomp.py                | ?                                                                           |
| tencentcloud-sdk-python | ?                                                                           |
| texttable               | ?                                                                           |
| twilio                  | ?                                                                           |
| tzlocal                 | ?                                                                           |

## memo

- tzlocal
  tzlocal is fixed at 2.1 until apscheduler major version upgrade

- croniter
  It uses datetime.utcnow and datetime.utcfromtimestamp which are deprecated in Python 3.12.<br>
  [Deprecate utcnow and utcfromtimestamp #103857](https://github.com/python/cpython/issues/103857)

- statsd-tags
  The github page on the pypi site is broken. Development finished?<br>
  [statsd-tags](https://pypi.org/project/statsd-tags/)

---

## requirements-dev.txt

| name             | python 3.12 |
| ---------------- | ----------- |
| pylint           | 〇(3.0.0～)           |
| pytest           | 〇(7.3.2～)           |
| pytest-cov       | △(Waiting for pull request merge)           |
| pytest-xdist     | ?           |
| setuptools       | ?           |
| sphinx           | ?           |
| sphinx_rtd_theme | ?           |
| tox              | 〇           |

## memo

- sphinx_rtd_theme
  Since sphinx 7.2 is being supported, it cannot be updated to sphinx 7.2 until the support is completed.

**Comment by nsano-rururu** [comment]
Python 3.12 will be officially released on October 2nd, 2023

**Reply by nsano-rururu** [reply]
If croniter and statsd-tags stop working after the upcoming new Python 3.12 release, you may want to consider switching to another library.

**Comment by jertel** [comment]
aiohttp is currently broken with Python 3.12. See https://github.com/aio-libs/aiohttp/issues/7229 for more information.

I'm pinning the Dockerfiles to 3.11 until 3.12 matures.

**Reply by nsano-rururu** [reply]
python 3.12 
```sh
============================================================================================= short test summary info ==============================================================================================
FAILED base_test.py::test_match - AttributeError: 'called_with' is not a valid assertion. Use a spec for the mock if 'called_with' is meant to be an attribute.
FAILED base_test.py::test_match_with_module - AttributeError: 'called_with' is not a valid assertion. Use a spec for the mock if 'called_with' is meant to be an attribute.
FAILED base_test.py::test_agg_matchtime_timestamp_field - AttributeError: 'called_with' is not a valid assertion. Use a spec for the mock if 'called_with' is meant to be an attribute.
FAILED base_test.py::test_agg_matchtime - AttributeError: 'called_with' is not a valid assertion. Use a spec for the mock if 'called_with' is meant to be an attribute.
FAILED alerters/command_test.py::test_command_getinfo - AttributeError: 'called_with' is not a valid assertion. Use a spec for the mock if 'called_with' is meant to be an attribute.
FAILED alerters/command_test.py::test_command_old_style_string_format1 - AttributeError: 'called_with' is not a valid assertion. Use a spec for the mock if 'called_with' is meant to be an attribute.
FAILED alerters/command_test.py::test_command_old_style_string_format2 - AttributeError: 'called_with' is not a valid assertion. Use a spec for the mock if 'called_with' is meant to be an attribute.
FAILED alerters/command_test.py::test_command_pipe_match_json - AttributeError: 'called_with' is not a valid assertion. Use a spec for the mock if 'called_with' is meant to be an attribute.
FAILED alerters/command_test.py::test_command_pipe_alert_text - AttributeError: 'called_with' is not a valid assertion. Use a spec for the mock if 'called_with' is meant to be an attribute.
FAILED alerters/command_test.py::test_command_fail_on_non_zero_exit - AttributeError: 'called_with' is not a valid assertion. Use a spec for the mock if 'called_with' is meant to be an attribute.
FAILED base_test.py::test_agg_with_aggregation_key - AttributeError: 'called_with' is not a valid assertion. Use a spec for the mock if 'called_with' is meant to be an attribute.
FAILED alerters/gelf_test.py::test_gelf_sent_tcp_with_custom_ca - AttributeError: <module 'ssl' from '/usr/local/lib/python3.12/ssl.py'> does not have the attribute 'wrap_socket'
======================================================================= 12 failed, 958 passed, 4 skipped, 306 warnings in 331.60s (0:05:31) ========================================================================

```
[Prevent prefix "called_" for methods on mock objects in safe mode #100690](https://github.com/python/cpython/issues/100690)
[[Python 3.12] AttributeError: module 'ssl' has no attribute 'wrap_socket' #795](https://github.com/eventlet/eventlet/issues/795)


https://projects.gentoo.org/python/guide/porting.html#called-with-and-other-invalid-assertions-now-trigger-an-error
